runtime/src/alloc.rs

```diff
@@ -0,0 +1,68 @@
+// Copyright 2023 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Minimum buffer between heap and stack
+const MEMORY_GAP: usize = 0x1000;
+
+#[allow(clippy::missing_safety_doc)]
+#[no_mangle]
+pub unsafe extern "C" fn sys_alloc_aligned(bytes: usize, align: usize) -> *mut u8 {
+    #[cfg(target_arch = "riscv32")]
+    extern "C" {
+        // https://lld.llvm.org/ELF/linker_script.html#sections-command
+        static _end: u8;
+    }
+
+    // Pointer to next heap address to use, or 0 if the heap has not yet been
+    // initialized.
+    static mut HEAP_POS: usize = 0;
+
+    // SAFETY: Single threaded, so nothing else can touch this while we're working.
+    let mut heap_pos = HEAP_POS;
```

Comment:
Why not just use HEAP_POS directly?

Reply:
This pattern is copied from RISCV Zero. I don't know why. 
From what I see, the pattern here is "update HEAP_POS when everything is right and correct, otherwise leave it as is".

Reply:
I think this will make sense once the allocator starts returning null or zero or something in case of failure. In those failure cases we don't want to modify HEAP_POS.